### PR TITLE
Add automated maintanence and cleanup via clang-tools to makefile

### PR DESCRIPTION
**Pull request by SteveBronder**
The clang-tidy and clang-format tools have been added to the
make file so that it's easier to keep all of the code consistent

#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

This adds

1. make methods for `clang-format` and `clang-tidy` so that it's easy to run them over the files in source. clang-format uses the same `.clang-format` file used in [the math library](https://github.com/SteveBronder/math/blob/develop/.clang-format)

There are only two actual places I made changes in the code

1. in [` src/stan/lang/generator/generate_fun_inst_templ_params.hpp`](https://github.com/stan-dev/stan/compare/develop...SteveBronder:cleanup/clang-format?expand=1#diff-6d4c39775589e2beeb3f9528bd0c0251R1) I removed the macro newline `\` and put a `// NOLINT` at the top
2. in [`src/stan/callbacks/stream_writer.hpp`](https://github.com/stan-dev/stan/compare/develop...SteveBronder:cleanup/clang-format?expand=1#diff-d9a5f54d53833b98b437bbfbb30158c9R26) I made the constructor on line 26 `explicit` since it _can_ accept only one argument

```cpp
stream_writer(std::ostream& output,
                         const std::string& comment_prefix = "")
```

#### Intended Effect

There should be no change to how code is executed. The tools are just so that the stan and math library have similar C++ styles. If that's not a good thing for the code at this level then we can close this PR (it took me like 10 mins to run this so no biggie)

#### How to Verify

From develop you can run `make clang-format` to get the code generated here (minus the two places listed above

#### Side Effects

If there are side effects beyond the two mentioned above then that's bad

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses: Steve Bronder

### CODE REVIEW

**Comment by SteveBronder** [comment]
Who should review this?

**Comment by syclik** [comment]
Cool. I took a look at the makefile stuff and it looked good. I spot-checked a handful of the changes and they look good too. 

I could approve, but I think we'd want either @serban-nicusor-toptal or @seantalts to sign off on the Jenkins change?

**Comment by SteveBronder** [comment]
@seantalts and I did chat about the automatic formatting the other day and he was sort of against it. I'm fine one way or another with add/removing the Jenkins stuff

**Comment by seantalts** [comment]
I'm heavily pro automatic formatting (and tidying if there's stuff we always want to do, which it seems like there is!). I just think it's more standard and better for devs to have the tools installed locally with the git pre-commit hook in general and not rely on the server correcting after the fact, because that way the git history is also formatted correctly. Having a server mode is nice because you might have contributors who are aren't in it for the long haul and it's not worth setting up all the tooling necessarily, but it'd be nice if that didn't also encourage long time devs to not code to the standard and rely on the server to fix it. I'm fine with adding the server mode here since we have it in Math - once we move to monorepo we'll have it everywhere as well. I'll let @serban-nicusor-toptal review :)

**Comment by SteveBronder** [comment]
My b Sean I think I mis-characterised your opinion a bit there. 

I can make separate PRs for math and this repo to add git hooks to auto run clang-format  before a commit.

BUT(!!). In exchange I want us to move up to clang-format-6 (or even 7) because 5 is not available for my computer (Ubuntu 19.04) :)

Pretty sure 6 is installable via `brew` on mac and has been backported to 19.04

**Comment by SteveBronder** [comment]
Ty! I'm not authorized to merge the PR but if someone wants to slam that button for me then go for it!

**Comment by serban-nicusor-toptal** [comment]
@SteveBronder I have reverted the merge.

File: Jenkinsfile [hunk 1]
```diff
@@ -104,6 +104,53 @@ pipeline {
...
+            agent any
+            steps {
+                sh "printenv"
+                deleteDir()
+                retry(3) { checkout scm }
+                withCredentials([usernamePassword(credentialsId: 'a630aebc-6861-4e69-b497-fd7f496ec46b',
+                    usernameVariable: 'GIT_USERNAME', passwordVariable: 'GIT_PASSWORD')]) {
+                    sh """#!/bin/bash
+                        set -x
+                        git checkout -b ${branchName()}
```
Comment (serban-nicusor-toptal) [inline Jenkinsfile]: ` ${branchName()}` should be `${GIT_BRANCH}` if you're looking for the current one checked out by Jenkins.

File: Jenkinsfile [hunk 2]
```diff
@@ -104,6 +104,53 @@ pipeline {
...
+                            git config --global user.email "mc.stanislaw@gmail.com"
+                            git config --global user.name "Stan Jenkins"
+                            git add stan test
+                            git commit -m "[Jenkins] auto-formatting by `clang-format --version`"
+                            git push https://${GIT_USERNAME}:${GIT_PASSWORD}@github.com/${fork()}/math.git ${branchName()}
+                            echo "Exiting build because clang-format found changes."
+                            echo "Those changes are now found on stan-dev/stan under branch ${branchName()}"
+                            echo "Please 'git pull' before continuing to develop."
+                            exit 1
+                        fi"""
```
Comment (serban-nicusor-toptal) [inline Jenkinsfile]: This is just a visibility thing, but I would end the sh after fi.
```
  echo "Please 'git pull' before continuing to develop."
  exit 1
fi
"""
```

**Comment by bob-carpenter** [comment]
How do I install the right tools to not have the server do the formatting?  Should I just wait until the dust settles on this PR and wait for new instructions?  I thought I had clang-format plugged into emacs, but maybe the standards changed or something.  Do I need some kind of GitHub hook or something?

**Comment by SteveBronder** [comment]
If you are on a mac you should be able to do 

```bash
brew install clang-tools
```

I need to check what version that is for mac. The server version is 5 which is v old and should be updated to at least 6

**Comment by seantalts** [comment]
Wait don't do that!

**Comment by seantalts** [comment]
Sorry, didn't want Bob to run that yet and then have to clean it up after. 

There are instructions here: https://github.com/stan-dev/stan/wiki/Coding-Style-and-Idioms#clang-format

relevant line: 
`brew install https://raw.githubusercontent.com/Homebrew/homebrew-core/5d29e8a6bfbe2e7d8f9525645bb0c00879d0984a/Formula/clang-format.rb
`

**Comment by SteveBronder** [comment]
Oh!? Is that for the specific version 5? Apologies

**Comment by seantalts** [comment]
yeah, that's the specific version we use. My desire for the ideal situation here is seriously undermined by 1) brew not doing versioning in a reasonable way and 2) clang-format having different versions labeled as the same version number available and 3) clang-format not being backwards compatible.

**Comment by syclik** [comment]
@seantalts, you deserve a medal for figuring that out.

**Comment by bob-carpenter** [comment]
Do I need to run as superuser?  It errors out on permissions after scolding
me for trying to install from a URL.  I'll append full run (less the
huge list of new formulas).

- Bob

~/temp2$ brew install https://raw.githubusercontent.com/Homebrew/homebrew-core/5d29e8a6bfbe2e7d8f9525645bb0c00879d0984a/Formula/clang-format.rb
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 2 taps (homebrew/core and homebrew/cask).
==> New Formulae
aliyun-cli                     deno                           kubeaudit                      proper

...

ftimes                   lysp                     plan9port                tlsdate
go@1.4                   minisat                  postgres-xc              tmux-cssh

==> Consider using `brew extract clang-format ...`!
This will extract your desired clang-format version to a stable tap instead of
installing from an unstable URL!

######################################################################## 100.0%
Error: clang-format 2017-06-22 is already installed
To upgrade to 2017-11-14, run `brew upgrade clang-format`.
==> `brew cleanup` has not been run in 30 days, running now...
Removing: /Users/carp/Library/Caches/Homebrew/isl--0.20.high_sierra.bottle.tar.gz... (1.4MB)
Removing: /Users/carp/Library/Caches/Homebrew/libmpc--1.1.0.high_sierra.bottle.tar.gz... (114.2KB)
Removing: /Users/carp/Library/Caches/Homebrew/mpfr--4.0.1.high_sierra.bottle.tar.gz... (1.1MB)
Removing: /Users/carp/Library/Caches/Homebrew/ocaml--4.07.1.high_sierra.bottle.tar.gz... (85.0MB)
Removing: /Users/carp/Library/Caches/Homebrew/pcre--8.43.high_sierra.bottle.tar.gz... (1.8MB)
Removing: /Users/carp/Library/Caches/Homebrew/pkg-config--0.29.2.high_sierra.bottle.tar.gz... (235.8KB)
Removing: /Users/carp/Library/Caches/Homebrew/linkage.json... (13KB)
Removing: /Users/carp/Library/Caches/Homebrew/portable-ruby-2.3.7.leopard_64.bottle.tar.gz... (12.4MB)
Removing: /Users/carp/Library/Logs/Homebrew/pkg-config... (64B)
Removing: /Users/carp/Library/Logs/Homebrew/pcre... (64B)
Error: Permission denied @ unlink_internal - /usr/local/share/ghostscript/9.05/Resource/Font/logo10.pfb

**Comment by SteveBronder** [comment]
What version do you get if you do

```
clang-format --version
```

**Comment by seantalts** [comment]
The advice in the error message suggests trying replace the "install" verb
with "upgrade" in the original command (with the git url), what happens if
you try that? Or possibly just uninstalling the version you have first.

**Pull request merged**